src/compiler/checker.ts

```diff
@@ -15349,7 +15349,7 @@ namespace ts {
                 source.flags & TypeFlags.InstantiableNonPrimitive ? isTypeDerivedFrom(getBaseConstraintOfType(source) || unknownType, target) :
                 target === globalObjectType ? !!(source.flags & (TypeFlags.Object | TypeFlags.NonPrimitive)) :
                 target === globalFunctionType ? !!(source.flags & TypeFlags.Object) && isFunctionObjectType(source as ObjectType) :
-                hasBaseType(source, getTargetType(target));
+                hasBaseType(source, getTargetType(target)) || (isArrayType(target) && !isReadonlyArrayType(target) && isTypeDerivedFrom(source, globalReadonlyArrayType));

```

Comment:
would it make sense to instead write

```
isArrayType(target) && !isReadonlyArrayType(target) ? isTypeDerivedFrom(source, globalReadonlyArrayType) 
: hasBaseType(source, getTargetType(target))
```

I can't quite tell if the base type fallback is important for readonly arrays, but I'm guessing not.

Reply:
The base type check is handling anything explicitly extending from `ReadonlyArray`, so it's probably important.